alf/networks/preprocessors.py

```diff
@@ -129,3 +129,161 @@ def _preprocess(self, tensor):
         ret = self._embedding_net(tensor)
         # EncodingNetwork returns a pair
         return (ret if self._input_tensor_spec.is_discrete else ret[0])
+
+
+@gin.configurable
+class SharedEmbeddingPreprocessor(InputPreprocessor, SingletonInstanceNetwork):
+    """An embedding preprocessor that can be shared by multiple networks.
+    """
+
+    def __init__(self,
+                 input_tensor_spec,
+                 embedding_dim,
+                 conv_layer_params=None,
+                 fc_layer_params=None,
+                 activation=torch.relu_,
+                 last_activation=math_ops.identity,
+                 name="EmbeddingPreproc"):
+        """
+        Args:
+            input_tensor_spec (TensorSpec): the input spec
+            embedding_dim (int): output embedding size
+            conv_layer_params (tuple[tuple]): a tuple of tuples where each
+                tuple takes a format ``(filters, kernel_size, strides, padding)``,
+                where ``padding`` is optional.
+            fc_layer_params (tuple[int]): a tuple of integers representing FC
+                layer sizes.
+            activation (torch.nn.functional): activation applied to the embedding
+            last_activation (nn.functional): activation function of the
+                last layer specified by embedding_dim. ``math_ops.identity`` is
+                used by default.
+            name (str):
+        """
+        super().__init__(input_tensor_spec, name)
+        if input_tensor_spec.is_discrete:
+            assert isinstance(input_tensor_spec, BoundedTensorSpec)
+            N = input_tensor_spec.maximum - input_tensor_spec.minimum + 1
+            # use nn.Embedding to support a large dictionary
+            self._embedding_net = nn.Embedding(N, embedding_dim)
+        else:
+            # Only use an MLP for embedding a continuous input
+            self._embedding_net = alf.networks.EncodingNetwork(
+                input_tensor_spec=input_tensor_spec,
+                conv_layer_params=conv_layer_params,
+                fc_layer_params=fc_layer_params,
+                activation=activation,
+                last_layer_size=embedding_dim,
+                last_activation=last_activation)
+
+    def _preprocess(self, tensor):
+        assert get_outer_rank(tensor, self._input_tensor_spec) == 1, \
+            "Only supports one outer rank (batch dim)!"
+        ret = self._embedding_net(tensor)
+        # EncodingNetwork returns a pair
+        return (ret if self._input_tensor_spec.is_discrete else ret[0])
+
+
+class PreprocessorNetwork(Network):
```

Comment:
It's not proper to move PreprocessorNetwork here in preprocessors.py. PreprocessorNetwork is a network that has preprocessors, but itself is not a preprocessor.

Reply:
Putting ```PreprocessorNetwork``` in ``network.py``` does not seem right as there will be circular dependency.

Reply:
Try not doing `from .. .import`. instead just do `alf.networks....`

Reply:
Now move ```PreprocessorNetwork``` out of ```preprocessors.py``` to a separate file.